README.md

## Features

1. Simple to use!

## Getting started

Install the and import class file to use package.

## Usage

```dart
//plz wait.. example will be show
```

## Additional information

Comming.



CHANGELOG.md
## 0.0.5
* New Required parameter that named of selectedIndex

## 0.0.4
* Update Documents

## 0.0.3
* Intergrate All classes in flutter_buttom_floating_bar.dart

## 0.0.2
* Change the package name and Git repo name

## 0.0.1
* Initial release.





lib/floating_navigation_bar.dart
import 'package:flutter/material.dart';
import 'floating_navigation_item.dart';

class FloatingNavigationBar extends StatefulWidget {
  ///It is items what you want to add
  List<FloatingNavigationItem> items;

  ///You can take what is current index when you tap navigation item
  Function(int index) onTap;

  ///A double that determines how far it is from the bottom
  double bottomMargin;

  ///Determines how much evocation effect you want to add.
  double elevation;

  ///Decide the padding you want to put inside of NavigationBar
  EdgeInsets padding;

  ///Decide how much flexion you want to give to the corner.
  BorderRadius? borderRadius;

  ///Set up how fast movement of Cursor
  Duration duration;

  ///The Curve of item movement
  Curve? curves;

  ///You can change the navigationbar color, but default color follows at the 'ThemeData.ColorScheme.SurfaceColor'
  Color? backgroundColor;

  FloatingNavigationBar(
      {required this.items,
        required this.onTap,
        Key? key,
        this.bottomMargin = 15,
        this.elevation = 5,
        this.padding = const EdgeInsets.only(right: 14, left: 14, top: 10, bottom: 2),
        this.duration = const Duration(milliseconds: 300),
        this.curves,
        this.borderRadius,
        this.backgroundColor})
      : super(key: key);

  @override
  State<FloatingNavigationBar> createState() => _FloatingNavigationBarState();
}

class _FloatingNavigationBarState extends State<FloatingNavigationBar> with TickerProviderStateMixin {
  int _selectedIndex = 0;
  final double _itemWidth = 40;
  final double _cursorWidth = 5;

  initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    ThemeData themeData = Theme.of(context);

    List<Widget> navigationItems = [];
    for (int i = 0; i < widget.items.length; i++) {
      navigationItems.add(_getNavigationItem(i, widget.items[i], widget.onTap));
    }

    return Container(
      margin: EdgeInsets.only(bottom: widget.bottomMargin),
      child: Row(
        mainAxisSize: MainAxisSize.max,
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          Material(
            borderRadius: widget.borderRadius ?? BorderRadius.circular(30),
            elevation: widget.elevation,
            color: Colors.transparent,
            child: Container(
              padding: widget.padding,
              decoration: BoxDecoration(
                borderRadius: widget.borderRadius ?? BorderRadius.circular(30),
                color: widget.backgroundColor ?? themeData.colorScheme.surface,
                // boxShadow: [BoxShadow(color: themeData.colorScheme.shadow, blurRadius: 10,spreadRadius: 1, offset: const Offset(0, 4))]
              ),
              child: Column(
                mainAxisSize: MainAxisSize.min,
                children: [
                  Container(
                    height: 5,
                    width: _itemWidth*navigationItems.length,
                    child: Stack(
                      children: [
                        AnimatedPositioned(
                            left: _getCurrentCursorPosition(_itemWidth, _cursorWidth, navigationItems.length, _selectedIndex),
                            top: 0,
                            bottom: 0,
                            duration: Duration(milliseconds: 200),
                            curve: widget.curves??Curves.easeInOut,
                            child: Container(
                              width: _cursorWidth,
                              decoration: BoxDecoration(color: themeData.colorScheme.primary, borderRadius: BorderRadius.circular(10)),
                            )),
                      ],
                    ),
                  ),
                  Row(
                    mainAxisSize: MainAxisSize.min,
                    children: navigationItems,
                  ),
                ],
              ),
            ),
          ),
        ],
      ),
    );
  }

  Widget _getNavigationItem(int index, FloatingNavigationItem item, Function(int index) onTap) {
    AnimationController animation = AnimationController(vsync: this);
    ThemeData themeData = Theme.of(context);
    Color? itemColor = _selectedIndex == index ? themeData.colorScheme.primary : null;
    return Container(
      width: _itemWidth,
      height: _itemWidth,
      child: Material(
          color: Colors.transparent,
          child: InkWell(
            customBorder: const CircleBorder(),
            child: Icon(
              item.icon,
              color: itemColor,
            ),
            onTap: () {
              _selectedIndex = index;
              onTap(index);
              setState(() {});
            },
          )),
    );
  }

  double _getCurrentCursorPosition(double itemWidth, double cursorWidth, int itemCount, int currentIndex) {
    final double defaultPosition = itemWidth / 2 - cursorWidth/1.8;
    double currentPosition = 0;
    if (currentIndex <= itemCount) {
      currentPosition = defaultPosition + (itemWidth * currentIndex);
    }
    return currentPosition;
  }
}


lib/flutter_bottom_floating_bar.dart
import 'package:flutter/material.dart';

class FloatingNavigationBar extends StatefulWidget {
  ///It is items what you want to add
  List<FloatingNavigationItem> items;

  ///You can take what is current index when you tap navigation item
  Function(int index) onTap;

  ///A double that determines how far it is from the bottom
  double bottomMargin;

  ///Determines how much evocation effect you want to add.
  double elevation;

  ///Decide the padding you want to put inside of NavigationBar
  EdgeInsets padding;

  ///Decide how much flexion you want to give to the corner.
  BorderRadius? borderRadius;

  ///Set up how fast movement of Cursor
  Duration duration;

  ///The Curve of item movement
  Curve? curves;

  ///You can change the navigationbar color, but default color follows at the 'ThemeData.ColorScheme.SurfaceColor'
  Color? backgroundColor;

  FloatingNavigationBar(
      {required this.items,
        required this.onTap,
        Key? key,
        this.bottomMargin = 15,
        this.elevation = 5,
        this.padding = const EdgeInsets.only(right: 14, left: 14, top: 10, bottom: 2),
        this.duration = const Duration(milliseconds: 300),
        this.curves,
        this.borderRadius,
        this.backgroundColor})
      : super(key: key);

  @override
  State<FloatingNavigationBar> createState() => _FloatingNavigationBarState();
}

class _FloatingNavigationBarState extends State<FloatingNavigationBar> with TickerProviderStateMixin {
  int _selectedIndex = 0;
  final double _itemWidth = 40;
  final double _cursorWidth = 5;

  initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    ThemeData themeData = Theme.of(context);

    List<Widget> navigationItems = [];
    for (int i = 0; i < widget.items.length; i++) {
      navigationItems.add(_getNavigationItem(i, widget.items[i], widget.onTap));
    }

    return Container(
      margin: EdgeInsets.only(bottom: widget.bottomMargin),
      child: Row(
        mainAxisSize: MainAxisSize.max,
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          Material(
            borderRadius: widget.borderRadius ?? BorderRadius.circular(30),
            elevation: widget.elevation,
            color: Colors.transparent,
            child: Container(
              padding: widget.padding,
              decoration: BoxDecoration(
                borderRadius: widget.borderRadius ?? BorderRadius.circular(30),
                color: widget.backgroundColor ?? themeData.colorScheme.surface,
                // boxShadow: [BoxShadow(color: themeData.colorScheme.shadow, blurRadius: 10,spreadRadius: 1, offset: const Offset(0, 4))]
              ),
              child: Column(
                mainAxisSize: MainAxisSize.min,
                children: [
                  Container(
                    height: 5,
                    width: _itemWidth*navigationItems.length,
                    child: Stack(
                      children: [
                        AnimatedPositioned(
                            left: _getCurrentCursorPosition(_itemWidth, _cursorWidth, navigationItems.length, _selectedIndex),
                            top: 0,
                            bottom: 0,
                            duration: Duration(milliseconds: 200),
                            curve: widget.curves??Curves.easeInOut,
                            child: Container(
                              width: _cursorWidth,
                              decoration: BoxDecoration(color: themeData.colorScheme.primary, borderRadius: BorderRadius.circular(10)),
                            )),
                      ],
                    ),
                  ),
                  Row(
                    mainAxisSize: MainAxisSize.min,
                    children: navigationItems,
                  ),
                ],
              ),
            ),
          ),
        ],
      ),
    );
  }

  Widget _getNavigationItem(int index, FloatingNavigationItem item, Function(int index) onTap) {
    AnimationController animation = AnimationController(vsync: this);
    ThemeData themeData = Theme.of(context);
    Color? itemColor = _selectedIndex == index ? themeData.colorScheme.primary : null;
    return Container(
      width: _itemWidth,
      height: _itemWidth,
      child: Material(
          color: Colors.transparent,
          child: InkWell(
            customBorder: const CircleBorder(),
            child: Icon(
              item.icon,
              color: itemColor,
            ),
            onTap: () {
              _selectedIndex = index;
              onTap(index);
              setState(() {});
            },
          )),
    );
  }

  double _getCurrentCursorPosition(double itemWidth, double cursorWidth, int itemCount, int currentIndex) {
    final double defaultPosition = itemWidth / 2 - cursorWidth/1.8;
    double currentPosition = 0;
    if (currentIndex <= itemCount) {
      currentPosition = defaultPosition + (itemWidth * currentIndex);
    }
    return currentPosition;
  }
}



class FloatingNavigationItem {
  IconData icon;


  FloatingNavigationItem({
    required this.icon,
  });
}

